### PR TITLE
Migrate to new authentication infrastructure

### PR DESCRIPTION
**Pull request by benjumanji**
- Move away from auth0 prop grant to oauth2 resource owner password grant.
- Point at new hypervolt authentication infrastructure.
- Audience is not required, as it is mapped based on the client.

### CODE REVIEW

File: custom_components/hypervolt_charger/hypervolt_api_client.py [hunk 1]
```diff
@@ -75,12 +75,10 @@ async def login(self, session: aiohttp.ClientSession) -> str:
             session.headers["user-agent"] = self.get_user_agent()
 
             async with session.post(
-                "https://auth.hypervolt.co.uk/oauth/token",
+                "https://kc.hypervolt.co.uk/realms/retail-customers/protocol/openid-connect/token",
```
Comment (gndean) [inline custom_components/hypervolt_charger/hypervolt_api_client.py]: This raises
```'Cannot connect to host kc.hypervolt.co.uk:443 ssl:default [Name or service not known]'```

If I `dig kc.hypervolt.co.uk` I don't get a response.

File: custom_components/hypervolt_charger/hypervolt_api_client.py [hunk 2]
```diff
@@ -75,12 +75,10 @@ async def login(self, session: aiohttp.ClientSession) -> str:
...
                 data={
-                    "audience": "https://api.hypervolt.co.uk",
-                    "client_id": "1BfAeDNKfu7mfPWCm6XTsZeM2QYhhju2",  # Mimic mobile app. If Hypervolt are reading this: Please give the HA community our own Client ID and allow us to support the HA OAuth2 login flow
-                    "grant_type": "http://auth0.com/oauth/grant-type/password-realm",
-                    "realm": "Username-Password-Authentication",
+                    "client_id": "home-assistant",  # Mimic mobile app. If Hypervolt are reading this: Please give the HA community our own Client ID and allow us to support the HA OAuth2 login flow
```
Comment (gndean) [inline custom_components/hypervolt_charger/hypervolt_api_client.py:80]: This comment can be dropped now 😄

File: custom_components/hypervolt_charger/hypervolt_api_client.py (hunk 1 above)
Comment (gndean) [reply custom_components/hypervolt_charger/hypervolt_api_client.py]: I tried:
`kb.hypervolt.co.uk` -> gives 403
`kc.prod.hypervolt.co.uk` -> gives 400
🤷

**Review by gndean — CHANGES_REQUESTED** [review]
The new auth domain isn't working for me unfortunately.

**Comment by benjumanji** [comment]
@gndean can you do me a favour: I am not by any means a python expert. I would have expected, based on my brief reading of https://docs.aiohttp.org/en/stable/client_quickstart.html#more-complicated-post-requests that this auth request would be sending _form encoded_ data. If it isn't that would explain the 400.

**Comment by gndean** [comment]
Yes, it's form encoded. Here's what the raw HTTP request looks like:

```POST /oauth/token HTTP/1.1
Host: auth.hypervolt.co.uk
user-agent: home-assistant-hypervolt-charger/2.0.1
Accept: */*
Accept-Encoding: gzip, deflate
Content-Length: 370
Content-Type: application/x-www-form-urlencoded

audience=https%3A%2F%2Fapi.hypervolt.co.uk&client_id=1BfAeDNKfu7mfPWCm6XTsZeM2QYhhju2&grant_type=http%3A%2F%2Fauth0.com%2Foauth%2Fgrant-type%2Fpassword-realm&realm=Username-Password-Authentication&scope=openid+profile+email+offline_access&username=...```

**Comment by benjumanji** [comment]
So I would hope that the new request then would roughly correspond with

```
❯ curl --request POST 'https://kc.prod.hypervolt.co.uk/realms/retail-customers/protocol/openid-connect/token' \
      --header 'Content-Type: application/x-www-form-urlencoded' \
      --data-urlencode 'client_id=native-with-token-exchange' \
      --data-urlencode 'password=XXX' --data-urlencode 'username=XXX' \
      --data-urlencode 'grant_type=password'
{"access_token":"XXX","expires_in":3600,"refresh_expires_in":5400,"refresh_token":"XXX","token_type":"Bearer","not-before-policy":1711661767,"session_state":"XXX","scope":"profile social email"}
```

I mean feel free to test with your own username and password. Maybe I am missing some url encoding of the params?

**Comment by gndean** [comment]
OK, so using the domain `kc.prod.hypervolt.co.uk`, not `kc.hypervolt.co.uk`, as in this PR, is working.
I did guess this and thought I had tried it previously and received a 400 but I must have changed something else at the same time.

Also, should the `client_id` be `"home-assistant"` or `"native-with-token-exchange"`?

Can you update the PR to include this correct domain name please. Then I can retest and hopefully merge.

**Review by gndean — APPROVED** [review]

**Pull request merged**